### PR TITLE
Basic support for `match_first` declarations.

### PR DESCRIPTION
**Pull request by zygoloid**
Add basic support for `match_first` declarations to explorer, as described in https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#prioritization-rule. The concrete syntax here is not yet approved, but the semantics are necessary to implement other approved features in explorer so we're intentionally getting a little ahead of the design here.

### CODE REVIEW

File: explorer/syntax/parser.ypp
Comment (jonmeow) [inline explorer/syntax/parser.ypp:1229]: Do you mean to support non-external declarations in match_first? You might want to handle this a little differently, because impl_kind_intro here allows it. You might want to add a test for it, to ensure it works correctly (e.g., if not supported, a compile error).

File: explorer/interpreter/impl_scope.cpp
Comment (jonmeow) [inline explorer/interpreter/impl_scope.cpp:243]: I'll note, a consequence of this particular syntax choice is that an external impl cannot be disambiguated in more than one priority ordering. i.e., if there is any ambiguity, it must be provide priorities for _all_ ambiguities. I'm not sure that the "declared more than once" comment addresses that issue, since couldn't an external impl be prioritized versus an impl in a different, imported package? (note, I'm particularly uncertain on the last since I know there are planned restrictions, but I'm not sure)

The flipside is, syntax-wise, I'm not sure how one could refer to another package's external impls directly in order to establish an ordering.

File: explorer/syntax/lexer.lpp
Comment (jonmeow) [inline explorer/syntax/lexer.lpp]: How exploratory is this? Would it make sense to tag this as `__match_first`, similar to what we've done for other experimental features (like `__mix`)? Or is this more final?

**Review by jonmeow — COMMENTED** [review]
I think the code is mostly fine, just a few comments on specific issues.

File: explorer/syntax/parser.ypp
Comment (zygoloid) [reply explorer/syntax/parser.ypp:1229]: `match_first` is only permitted at the top level, and in principle non-external declarations are only permitted within a class. But we don't actually enforce that `external` is used properly right now. I'll send a separate PR for that, then add tests here once it lands.

File: explorer/syntax/lexer.lpp
Comment (zygoloid) [reply explorer/syntax/lexer.lpp]: The keyword [appears in the design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#prioritization-rule) and was added there by an approved proposal, but it's in an "open questions" section. I don't have a strong view here. If you'd prefer that keywords that haven't been adopted by a proposal are all named with a `__` then I'll do that.

File: explorer/interpreter/impl_scope.cpp
Comment (zygoloid) [reply explorer/interpreter/impl_scope.cpp:243]: `impl` declarations in a `match_first` are "real" declarations of the `impl`, so they can only appear within the library that "owns" that `impl`. You can't use `match_first` to prioritize across libraries, but it wouldn't make sense to do so anyway because `match_first` only disambiguates between` impl`s with the same type structure, and there's at most one library in which an `impl` with a given type structure can appear. (Nearly all of this is not yet implemented; `match_first` syntax is just an early step into this `impl` prioritization system.)

File: explorer/syntax/lexer.lpp
Comment (jonmeow) [reply explorer/syntax/lexer.lpp]: Per discussion -- let's use `__match_first`, since `match_first` is in examples, but `match_first` is somewhat confusing given there's also `match` (which also matches the first) and may not be the final keyword name as a consequence.

File: explorer/syntax/parser.ypp
Comment (jonmeow) [reply explorer/syntax/parser.ypp:1229]: Separate PR works for me.

**Review by jonmeow — APPROVED** [review]
Approving to unblock merging once __match_first is done.

**Pull request merged**
